monai/networks/nets/dints.py

```diff
@@ -0,0 +1,556 @@
+                    nn.InstanceNorm3d(filter_nums[res_idx]),
+                    nn.ReLU(),
+                    nn.Conv3d(filter_nums[res_idx], filter_nums[res_idx + 1], 3, stride=2, padding=1, bias=False),
+                    nn.InstanceNorm3d(filter_nums[res_idx + 1]),
+                )
+                self.stem_up[str(res_idx)] = nn.Sequential(
+                    nn.ReLU(),
+                    nn.Conv3d(filter_nums[res_idx + 1], filter_nums[res_idx], 3, stride=1, padding=1, bias=False),
+                    nn.InstanceNorm3d(filter_nums[res_idx]),
+                    nn.Upsample(scale_factor=2, mode="trilinear", align_corners=True),
+                )
+
+            else:
+                self.stem_down[str(res_idx)] = nn.Sequential(
+                    nn.Upsample(scale_factor=1 / (2 ** res_idx), mode="trilinear", align_corners=True),
+                    nn.Conv3d(in_channels, filter_nums[res_idx], 3, stride=1, padding=1, bias=False),
+                    nn.InstanceNorm3d(filter_nums[res_idx]),
+                )
+                self.stem_up[str(res_idx)] = nn.Sequential(
+                    nn.ReLU(),
+                    nn.Conv3d(filter_nums[res_idx], filter_nums[res_idx], 3, stride=1, padding=1, bias=False),
+                    nn.InstanceNorm3d(filter_nums[res_idx]),
+                    nn.Conv3d(filter_nums[res_idx], num_classes, 1),
+                    nn.Upsample(scale_factor=2 ** res_idx, mode="trilinear", align_corners=True),
+                )
+
+        # define NAS search space
+        if code is None:
+            code_a = np.ones((num_blocks, len(code2out)))
+            code_c = np.ones((num_blocks, len(code2out), cell_ops))
+        else:
+            code_a = code[1]
+            code_c = F.one_hot(torch.from_numpy(code[2]), cell_ops).numpy()
+        self.cell_tree = nn.ModuleDict()
+        self.memory = np.zeros((num_blocks, len(code2out), cell_ops))
+        for blk_idx in range(num_blocks):
+            for res_idx in range(len(code2out)):
+                if code_a[blk_idx, res_idx] == 1:
+                    self.cell_tree[str((blk_idx, res_idx))] = cell(
+                        filter_nums[code2in[res_idx] + int(use_stem)],
+                        filter_nums[code2out[res_idx] + int(use_stem)],
+                        code2ops[res_idx],
+                        code_c[blk_idx, res_idx],
+                    )
+                    self.memory[blk_idx, res_idx] = np.array(
+                        [
+                            _.memory + self.cell_tree[str((blk_idx, res_idx))].preprocess.memory if _ is not None else 0
+                            for _ in self.cell_tree[str((blk_idx, res_idx))].op._ops[:cell_ops]
+                        ]
+                    )
+
+        # define cell and macro arhitecture probabilities
+        self.log_alpha_c = torch.nn.Parameter(
+            torch.zeros(num_blocks, len(code2out), cell_ops).normal_(1, 0.01).cuda().requires_grad_()
```

Comment:
the tensors are hard-coded to cuda device, I think this should be changed to be optional.